bootstrap: Consolidate coverage test suite steps into a single step

Zalathar opened the pull request:
Now that I have more understanding of bootstrap steps, and a renewed distaste for unnecessary macros, I have managed to express the subtleties of the `tests/coverage` test suite in a single step defined in ordinary code, with no need for helper macros.

Deciding which modes to run is still a bit clunky due to limitations in existing ShouldRun/PathSet APIs, but I think it's a net improvement over having to declare several different steps to handle the suite path and aliases.

The interaction with `--skip` isn't as nice as I'd like, but all of the known limitations are limitations that already existed in the previous implementation.

One minor change is that by default compiletest is now invoked in `coverage-run` mode even when cross-compiling. However, in that situation compiletest still knows that it should skip all of the individual coverage-run tests.

r? jieyouxu (or reassign)

Zalathar committed:
Consolidate coverage test suite steps into a single step

jieyouxu commented:
I'm going to reroll because I'm not super sure about the `PathSet` matching logic. r? bootstrap

Kobzol commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -828,3 +828,34 @@ fn test_test_compiler() {
+    ];
+
+    for &Case { cmd, expected } in cases {
+        println!("- {cmd:?}");
```
Leftover debugging print?

Zalathar replied:
Leaving it in was deliberate. If the test fails then this makes it possible to see which case failed, and if the test passes then the output is hidden by output capture.

Let me know if you still want it removed, but I think leaving it in is worthwhile.

Kobzol replied:
Ok, then let's leave it :)

Kobzol commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -1473,99 +1426,96 @@ impl Step for RunMake {
 pub struct Coverage {
     pub compiler: Compiler,
     pub target: TargetSelection,
+    pub(crate) mode: &'static str,
```
```suggestion
    pub mode: &'static str,
```
It's not super important, but the rest of the code here uses `pub`, so let's keep it unified, otherwise it looks like the visibility was different for some special reason.

Kobzol reviewed (commented):
This is a nice simplification, thanks.

Something really weird is going on with the `--skip` logic though.
In https://github.com/rust-lang/rust/blob/089bd510cf5ac835f1c3755f44fa2d15d211f09a/src/bootstrap/src/core/builder/mod.rs#L403, a step is skipped only if *all* of its pathsets and aliases are skipped, which seems quite suspicious to me, that makes some default steps really hard to skip.

On the other hand, the logic in [`ensure_if_default`](https://github.com/rust-lang/rust/blob/089bd510cf5ac835f1c3755f44fa2d15d211f09a/src/bootstrap/src/core/builder/mod.rs#L1506) skips the step if *any* of its paths is skipped, which makes much more sense to me.

At the very least, these two skip logics should probably be merged somehow.

In any case, this doesn't seem to make the status quo worse in this regard, as you stated in the PR description. I left two nits, otherwise you can r=me.

Zalathar commented:
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/089bd510cf5ac835f1c3755f44fa2d15d211f09a..5298f85a0a2eadad45368137d57c1ff5a4adfd92).

Kobzol commented:
Thanks! I'll approve once CI is green.

Kobzol commented:
@bors r+

bors commented:
:pushpin: Commit 5298f85a0a2eadad45368137d57c1ff5a4adfd92 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 5298f85a0a2eadad45368137d57c1ff5a4adfd92 -->
<!-- homu: {"type":"Approved","sha":"5298f85a0a2eadad45368137d57c1ff5a4adfd92","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
> , a step is skipped only if _all_ of its pathsets and aliases are skipped, which seems quite suspicious to me, that makes some default steps really hard to skip.
> On the other hand, the logic in [`ensure_if_default`](https://github.com/rust-lang/rust/blob/089bd510cf5ac835f1c3755f44fa2d15d211f09a/src/bootstrap/src/core/builder/mod.rs#L1506) skips the step if _any_ of its paths is skipped, which makes much more sense to me.
> 
> At the very least, these two skip logics should probably be merged somehow.

Note that a single step can have multiple paths representing different tools/crates, such that the different parts of the step can be run/skipped independently.

For example, the bootstrap tools `src/tools/jsondoclint` and `src/tools/suggest-tests` are tested by the same step (`test::CrateBootstrap`) via different paths. But passing `--skip jsondoclint` shouldn't cause the tests for `suggest-tests` to be skipped!
